File: caipiao-game/utils/history.js
/**
 * 历史开奖数据分析（移植自 Python history.py）
 *
 * 功能：
 *   - 加载历史数据
 *   - 计算单号频率 / 出现率 / 热冷号分类
 *   - 对一组号码做命中分析（平均命中数 / 命中分布 / 完全匹配期数）
 */

const ssqHistory = require("../data/ssq_history.js");
const dltHistory = require("../data/dlt_history.js");

const RULES = {
  ssq: {
    primaryRange: rangeArr(1, 33),
    primaryCount: 6,
    primaryLabel: "红球",
    secondaryRange: rangeArr(1, 16),
    secondaryCount: 1,
    secondaryLabel: "蓝球",
  },
  dlt: {
    primaryRange: rangeArr(1, 35),
    primaryCount: 5,
    primaryLabel: "前区",
    secondaryRange: rangeArr(1, 12),
    secondaryCount: 2,
    secondaryLabel: "后区",
  },
};

function rangeArr(from, to) {
  const a = [];
  for (let i = from; i <= to; i++) a.push(i);
  return a;
}

// ===== 数据加载 =====

/**
 * 解析紧凑格式："issue|date|primary|secondary"
 * 返回 [{ issue, date, primary: [n,...], secondary: [n,...] }, ...]
 */
function _parseHistory(lines, kind) {
  const out = [];
  for (const line of lines) {
    if (!line || typeof line !== "string") continue;
    const parts = line.split("|");
    if (parts.length !== 4) continue;
    const [issue, date, pStr, sStr] = parts;
    const primary = pStr.split(",").map(s => parseInt(s.trim(), 10)).filter(n => !isNaN(n));
    let secondary;
    if (kind === "ssq") {
      secondary = [parseInt(sStr.trim(), 10)];
    } else {
      secondary = sStr.split(",").map(s => parseInt(s.trim(), 10));
    }
    if (primary.length === 0 || secondary.length === 0) continue;
    out.push({ issue, date, primary, secondary });
  }
  return out;
}

function loadHistory(kind) {
  const lines = kind === "ssq" ? ssqHistory : dltHistory;
  return _parseHistory(lines, kind);
}

// ===== 单号频率统计 =====

function _frequency(draws, getter) {
  const counter = {};
  for (const d of draws) {
    for (const n of getter(d)) {
      counter[n] = (counter[n] || 0) + 1;
    }
  }
  return counter;
}

function _classifyLevels(stats) {
  if (stats.length === 0) return stats;
  const freqs = [...new Set(stats.map(s => s.frequency))].sort((a, b) => a - b);
  const n = freqs.length;
  let lowThreshold = freqs[0];
  let highThreshold = freqs[n - 1];
  if (n >= 3) {
    lowThreshold = freqs[Math.floor(n / 3) - 1];
    highThreshold = freqs[Math.min(n - 1, Math.floor(2 * n / 3))];
  }
  if (lowThreshold === highThreshold) lowThreshold = highThreshold - 1;
  for (const s of stats) {
    if (s.frequency >= highThreshold && highThreshold > lowThreshold) s.level = "hot";
    else if (s.frequency <= lowThreshold) s.level = "cold";
    else s.level = "warm";
  }
  return stats;
}

function _makeNumberStat(n, freq, total, poolSize) {
  const expectedRate = poolSize / total;
  const rate = total > 0 ? freq / total : 0;
  return { number: n, frequency: freq, rate, expectedRate, deviation: rate - expectedRate, level: "warm" };
}

function computeNumberStats(kind) {
  const draws = loadHistory(kind);
  const rule = RULES[kind];
  const total = Math.max(1, draws.length);

  const primaryCounter = _frequency(draws, d => d.primary);
  const secondaryCounter = _frequency(draws, d => d.secondary);

  const primaryStats = rule.primaryRange.map(n =>
    _makeNumberStat(n, primaryCounter[n] || 0, total, rule.primaryRange.length)
  );
  const secondaryStats = rule.secondaryRange.map(n =>
    _makeNumberStat(n, secondaryCounter[n] || 0, total, rule.secondaryRange.length)
  );

  _classifyLevels(primaryStats);
  _classifyLevels(secondaryStats);

  return { primaryStats, secondaryStats, totalDraws: total };
}

function getTopBottom(kind, n = 5) {
  const { primaryStats, secondaryStats, totalDraws } = computeNumberStats(kind);
  if (totalDraws === 0) {
    return { primaryHot: [], primaryCold: [], secondaryHot: [], secondaryCold: [], totalDraws: 0 };
  }
  const _top = (arr, reverse) =>
    arr.slice().sort((a, b) => reverse ? b.frequency - a.frequency : a.frequency - b.frequency).slice(0, n);
  return {
    primaryHot: _top(primaryStats, true),
    primaryCold: _top(primaryStats, false),
    secondaryHot: _top(secondaryStats, true),
    secondaryCold: _top(secondaryStats, false),
    totalDraws,
  };
}

// ===== 命中分析 =====

function _missingNumberStat(n, poolSize, total) {
  return {
    number: n, frequency: 0, rate: 0,
    expectedRate: total > 0 ? poolSize / total : 0,
    deviation: 0, level: "warm",
  };
}

function analyzeBet(kind, primaryNumbers, secondaryNumbers) {
  const { primaryStats, secondaryStats, totalDraws } = computeNumberStats(kind);
  const draws = loadHistory(kind);
  const rule = RULES[kind];

  const pSet = new Set(primaryNumbers);
  const sSet = new Set(secondaryNumbers);

  const pByNumber = {}; primaryStats.forEach(s => pByNumber[s.number] = s);
  const sByNumber = {}; secondaryStats.forEach(s => sByNumber[s.number] = s);

  let primaryHits = 0;
  let secondaryHits = 0;
  const primaryDist = {};
  const secondaryDist = {};
  const fullMatchIssues = [];

  for (const d of draws) {
    const pHit = d.primary.filter(n => pSet.has(n)).length;
    const sHit = d.secondary.filter(n => sSet.has(n)).length;
    primaryHits += pHit;
    secondaryHits += sHit;
    primaryDist[pHit] = (primaryDist[pHit] || 0) + 1;
    secondaryDist[sHit] = (secondaryDist[sHit] || 0) + 1;
    if (pHit === pSet.size && sHit === sSet.size) {
      fullMatchIssues.push(d.issue);
    }
  }

  const primaryAvg = totalDraws > 0 ? primaryHits / totalDraws : 0;
  const secondaryAvg = totalDraws > 0 ? secondaryHits / totalDraws : 0;

  const pSelected = primaryNumbers.map(n =>
    pByNumber[n] || _missingNumberStat(n, rule.primaryRange.length, totalDraws)
  );
  const sSelected = secondaryNumbers.map(n =>
    sByNumber[n] || _missingNumberStat(n, rule.secondaryRange.length, totalDraws)
  );

  const levelBreakdown = { hot: 0, warm: 0, cold: 0 };
  for (const s of [...pSelected, ...sSelected]) {
    levelBreakdown[s.level] = (levelBreakdown[s.level] || 0) + 1;
  }

  return {
    bet: kind === "ssq"
      ? { reds: primaryNumbers.slice().sort((a, c) => a - c), blue: secondaryNumbers[0] }
      : { fronts: primaryNumbers.slice().sort((a, c) => a - c), backs: secondaryNumbers.slice().sort((a, c) => a - c) },
    totalDraws,
    primaryStats: pSelected,
    secondaryStats: sSelected,
    primaryAvgHits: primaryAvg,
    secondaryAvgHits: secondaryAvg,
    primaryHitDistribution: primaryDist,
    secondaryHitDistribution: secondaryDist,
    fullMatchCount: fullMatchIssues.length,
    fullMatchIssues: fullMatchIssues.slice(0, 10),
    levelBreakdown,
  };
}

// ===== 期间计算（截止时间逻辑）=====
//
// 彩票规则：
//   SSQ（双色球）：每周日/二/四 21:30 开奖
//                 销售截止：开奖当日 20:00
//   DLT（大乐透）：每周一/三/六 20:30 开奖
//                 销售截止：开奖当日 19:00
//
// 业务规则：
//   截止时间前 → 生成"当期"号码（即下一次开奖的期号）
//   截止时间后 → 生成"下一期"号码（即再下一次开奖的期号）
//   这样确保：保存到彩票库的票，开奖后能正确对照
//
// 期间号规则（每年从 001/0001 重新计）：
//   SSQ: YYYYNNN  (如 2026088)
//   DLT: YYNNN    (如 26086)

const DRAW_RULES = {
  ssq: {
    drawDays: [0, 2, 4],   // 周日(0)、周二(2)、周四(4)
    drawHour: 21,
    drawMin:  30,
    cutoffHour: 20,
    cutoffMin:  0,
  },
  dlt: {
    drawDays: [1, 3, 6],   // 周一(1)、周三(3)、周六(6)
    drawHour: 20,
    drawMin:  30,
    cutoffHour: 19,
    cutoffMin:  0,
  },
};

/**
 * 找下一次开奖日（含今天若今天就是开奖日且未开奖）
 * @returns {Date} 开奖日（时分秒设为开奖时刻）
 */
function _nextDrawDate(lottery, now) {
  const r = DRAW_RULES[lottery];
  const candidate = new Date(now);
  candidate.setHours(r.drawHour, r.drawMin, 0, 0);
  if (r.drawDays.includes(candidate.getDay()) && now < candidate) {
    return candidate;
  }
  // 否则往后 1-7 天找
  for (let i = 1; i <= 7; i++) {
    const d = new Date(candidate);
    d.setDate(candidate.getDate() + i);
    if (r.drawDays.includes(d.getDay())) {
      d.setHours(r.drawHour, r.drawMin, 0, 0);
      return d;
    }
  }
  return null;
}

/**
 * 计算从基准日（含）到目标日（含）之间有多少个"开奖日"
 * 仅在同一年内有效
 */
function _countDrawDaysInYear(lottery, fromDate, toDate) {
  const r = DRAW_RULES[lottery];
  const yearStart = new Date(toDate.getFullYear(), 0, 1);
  let count = 0;
  const cur = new Date(yearStart);
  while (cur <= toDate) {
    if (r.drawDays.includes(cur.getDay())) count++;
    cur.setDate(cur.getDate() + 1);
  }
  return count;
}

/**
 * 根据当前时间 + 彩票类型 + 历史最新期，计算：
 *   - currentPeriod: "当期"期号（下一个开奖日）
 *   - targetPeriod:  "目标"期号（截止前=当期，截止后=下一期）
 *   - drawDate:      当期开奖日
 *   - targetDate:    目标期开奖日
 *   - isNextPeriod:  是否已过截止（true=生成的票是下一期）
 *   - daysUntilDraw: 距离当期开奖还有几天（0=今天）
 *   - hoursUntilCutoff: 距离当期截止还有几小时（负数=已截止）
 *
 * @param {'ssq'|'dlt'} lottery
 * @param {Date} [now] 不传则用 new Date()
 */
function getCurrentPeriod(lottery, now) {
  if (!now) now = new Date();
  const r = DRAW_RULES[lottery];
  if (!r) {
    return { error: `unknown lottery: ${lottery}` };
  }

  // 1) 找"当期"开奖日（下一次开奖）
  const drawDate = _nextDrawDate(lottery, now);

  // 2) 找"目标"开奖日（截止前=当期，截止后=下一期）
  const cutoff = new Date(drawDate);
  cutoff.setHours(r.cutoffHour, r.cutoffMin, 0, 0);
  let targetDate = drawDate;
  let isNextPeriod = false;
  if (now >= cutoff) {
    isNextPeriod = true;
    for (let i = 1; i <= 7; i++) {
      const d = new Date(drawDate);
      d.setDate(drawDate.getDate() + i);
      if (r.drawDays.includes(d.getDay())) {
        d.setHours(r.drawHour, r.drawMin, 0, 0);
        targetDate = d;
        break;
      }
    }
  }

  // 3) 计算当期 + 目标期号（基于历史最新期 + 推算）
  const draws = loadHistory(lottery);
  const latest = draws[draws.length - 1];
  const latestDate = new Date(latest.date);
  const latestIssue = latest.issue;
  const latestYear = lottery === 'ssq' ? parseInt(latestIssue.slice(0, 4)) : 2000 + parseInt(latestIssue.slice(0, 2));
  const latestNum = lottery === 'ssq' ? parseInt(latestIssue.slice(4)) : parseInt(latestIssue.slice(2));

  function _walk(fromDate, toDate) {
    // 从 fromDate（不含）到 toDate（含）的开奖日数
    const d = new Date(fromDate);
    d.setDate(d.getDate() + 1);
    let n = 0;
    while (d <= toDate) {
      if (r.drawDays.includes(d.getDay())) n++;
      d.setDate(d.getDate() + 1);
    }
    return n;
  }

  function _formatPeriod(year, num) {
    if (lottery === 'ssq') {
      return String(year) + String(num).padStart(3, '0');
    } else {
      return String(year).slice(-2) + String(num).padStart(3, '0');
    }
  }

  function _issueFor(targetD) {
    const tYear = targetD.getFullYear();
    if (tYear === latestYear) {
      const n = _walk(latestDate, targetD);
      return _formatPeriod(tYear, latestNum + n);
    } else {
      // 跨年了：当年从 1 开始
      const n = _countDrawDaysInYear(lottery, latestDate, targetD);
      return _formatPeriod(tYear, n);
    }
  }

  const currentPeriod = _issueFor(drawDate);
  const targetPeriod = _issueFor(targetDate);

  // 4) 倒计时
  const daysUntilDraw = Math.floor((drawDate - now) / 86400000);
  const hoursUntilCutoff = (cutoff - now) / 3600000;

  return {
    lottery,
    lotteryName: lottery === 'ssq' ? '双色球' : '大乐透',
    currentPeriod,    // 当期期号
    targetPeriod,     // 目标期号（截止后=下一期）
    drawDate,         // 当期开奖日
    targetDate,       // 目标期开奖日
    cutoff,           // 当期截止时间
    isNextPeriod,     // 是否已过截止
    isAfterDraw:      now >= drawDate,  // 当期是否已开奖
    daysUntilDraw,
    hoursUntilCutoff,
  };
}

// ===== 摘要 =====

function getSummary(kind) {
  const draws = loadHistory(kind);
  if (draws.length === 0) {
    return { lotteryType: kind, totalDraws: 0, earliestIssue: null, latestIssue: null, earliestDate: null, latestDate: null, hasData: false };
  }
  return {
    lotteryType: kind,
    totalDraws: draws.length,
    earliestIssue: draws[0].issue,
    latestIssue: draws[draws.length - 1].issue,
    earliestDate: draws[0].date,
    latestDate: draws[draws.length - 1].date,
    hasData: true,
  };
}

module.exports = {
  loadHistory,
  computeNumberStats,
  getTopBottom,
  analyzeBet,
  getSummary,
  getCurrentPeriod,
  RULES,
  DRAW_RULES,
};

File: history.py
"""
历史开奖数据分析模块。

提供：
- 从本地 JSON 加载历史开奖数据
- 单号频率 / 出现比率统计
- 热号 / 冷号 / 温号分类
- 对当前一组号码做"历史命中"分析
- JSON 合并 / 去重 / 排序

数据格式（data/{ssq,dlt}_history.json）：
{
  "lottery_type": "ssq" | "dlt",
  "last_update": "YYYY-MM-DD",
  "total_draws": int,
  "draws": [
    {"issue": str, "date": "YYYY-MM-DD",
     "reds"|"front": list[int],
     "blue": int (ssq) | "back": list[int] (dlt)}
  ]
}
"""

from __future__ import annotations

import json
from collections import Counter
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Iterable

# ===== 历史数据存储路径 =====
DATA_DIR = Path(__file__).parent / "data"
SSQ_HISTORY_FILE = DATA_DIR / "ssq_history.json"
DLT_HISTORY_FILE = DATA_DIR / "dlt_history.json"

# ===== 彩种规则 =====
SSQ_RULE = {
    "primary_range": range(1, 34),
    "primary_count": 6,
    "primary_label": "红球",
    "secondary_range": range(1, 17),
    "secondary_count": 1,
    "secondary_label": "蓝球",
}

DLT_RULE = {
    "primary_range": range(1, 36),
    "primary_count": 5,
    "primary_label": "前区",
    "secondary_range": range(1, 13),
    "secondary_count": 2,
    "secondary_label": "后区",
}

RULES = {
    "ssq": SSQ_RULE,
    "dlt": DLT_RULE,
}


def history_file(lottery_type: str) -> Path:
    return SSQ_HISTORY_FILE if lottery_type == "ssq" else DLT_HISTORY_FILE


# ===== 数据类 =====

@dataclass
class Draw:
    issue: str
    date: str
    reds: list[int] = field(default_factory=list)
    blue: int | None = None
    front: list[int] = field(default_factory=list)
    back: list[int] = field(default_factory=list)

    def primary(self) -> list[int]:
        return self.reds if self.reds else self.front

    def secondary(self) -> list[int]:
        if self.blue is not None:
            return [self.blue]
        return self.back

    def primary_set(self) -> set[int]:
        return set(self.primary())

    def secondary_set(self) -> set[int]:
        return set(self.secondary())


@dataclass
class HistorySummary:
    lottery_type: str
    total_draws: int
    last_update: str | None
    earliest_issue: str | None
    latest_issue: str | None
    earliest_date: str | None
    latest_date: str | None
    has_data: bool

    def to_dict(self) -> dict:
        return self.__dict__.copy()


@dataclass
class NumberStat:
    number: int
    frequency: int
    rate: float
    expected_rate: float
    deviation: float
    level: str

    def to_dict(self) -> dict:
        return {
            "number": self.number,
            "frequency": self.frequency,
            "rate": round(self.rate, 4),
            "expected_rate": round(self.expected_rate, 4),
            "deviation": round(self.deviation, 4),
            "level": self.level,
        }


@dataclass
class HitAnalysis:
    bet: dict
    total_draws: int
    primary_stats: list[dict]
    secondary_stats: list[dict]
    primary_avg_hits: float
    secondary_avg_hits: float
    primary_hit_distribution: dict
    secondary_hit_distribution: dict
    full_match_count: int
    full_match_issues: list
    level_breakdown: dict

    def to_dict(self) -> dict:
        return {
            "bet": self.bet,
            "total_draws": self.total_draws,
            "primary_stats": self.primary_stats,
            "secondary_stats": self.secondary_stats,
            "primary_avg_hits": round(self.primary_avg_hits, 4),
            "secondary_avg_hits": round(self.secondary_avg_hits, 4),
            "primary_hit_distribution": {str(k): v for k, v in self.primary_hit_distribution.items()},
            "secondary_hit_distribution": {str(k): v for k, v in self.secondary_hit_distribution.items()},
            "full_match_count": self.full_match_count,
            "full_match_issues": self.full_match_issues,
            "level_breakdown": self.level_breakdown,
        }


# ===== 加载 / 保存 =====

def _normalize_draw(raw: dict, lottery_type: str) -> Draw | None:
    issue = str(raw.get("issue", "")).strip()
    if not issue:
        return None

    draw_date = str(raw.get("date", "")).strip()

    if lottery_type == "ssq":
        reds = raw.get("reds") or raw.get("red") or []
        blue = raw.get("blue")
        if not reds:
            return None
        if isinstance(reds, str):
            reds = [int(x) for x in reds.replace(",", " ").split() if x.strip()]
        if not isinstance(blue, int):
            return None
        return Draw(issue=issue, date=draw_date, reds=[int(x) for x in reds], blue=int(blue))

    front = raw.get("front") or raw.get("fronts") or raw.get("reds") or []
    back = raw.get("back") or raw.get("backs") or []
    if not front or not back:
        return None
    if isinstance(front, str):
        front = [int(x) for x in front.replace(",", " ").split() if x.strip()]
    if isinstance(back, str):
        back = [int(x) for x in back.replace(",", " ").split() if x.strip()]
    return Draw(issue=issue, date=draw_date, front=[int(x) for x in front], back=[int(x) for x in back])


def load_history(lottery_type: str) -> list:
    path = history_file(lottery_type)
    if not path.exists():
        return []

    try:
        with path.open("r", encoding="utf-8") as f:
            payload = json.load(f)
    except (OSError, json.JSONDecodeError):
        return []

    raw_draws = payload.get("draws", []) if isinstance(payload, dict) else []
    draws = []
    for raw in raw_draws:
        draw = _normalize_draw(raw, lottery_type)
        if draw is not None:
            draws.append(draw)

    draws.sort(key=lambda d: d.issue)
    return draws


def save_history(lottery_type: str, draws: Iterable, last_update: str | None = None) -> Path:
    existing = load_history(lottery_type)
    by_issue = {d.issue: d for d in existing}
    for d in draws:
        by_issue[d.issue] = d

    merged = sorted(by_issue.values(), key=lambda d: d.issue)
    rule = RULES[lottery_type]

    if lottery_type == "ssq":
        serialized = [
            {"issue": d.issue, "date": d.date, "reds": sorted(d.reds), "blue": d.blue}
            for d in merged
        ]
    else:
        serialized = [
            {"issue": d.issue, "date": d.date, "front": sorted(d.front), "back": sorted(d.back)}
            for d in merged
        ]

    payload = {
        "lottery_type": lottery_type,
        "last_update": last_update or datetime.now().strftime("%Y-%m-%d"),
        "total_draws": len(merged),
        "primary_label": rule["primary_label"],
        "secondary_label": rule["secondary_label"],
        "primary_max": max(rule["primary_range"]),
        "secondary_max": max(rule["secondary_range"]),
        "draws": serialized,
    }

    path = history_file(lottery_type)
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as f:
        json.dump(payload, f, ensure_ascii=False, indent=2)
    return path


def get_summary(lottery_type: str) -> HistorySummary:
    draws = load_history(lottery_type)
    if not draws:
        return HistorySummary(
            lottery_type=lottery_type,
            total_draws=0,
            last_update=None,
            earliest_issue=None,
            latest_issue=None,
            earliest_date=None,
            latest_date=None,
            has_data=False,
        )

    return HistorySummary(
        lottery_type=lottery_type,
        total_draws=len(draws),
        last_update=None,
        earliest_issue=draws[0].issue,
        latest_issue=draws[-1].issue,
        earliest_date=draws[0].date,
        latest_date=draws[-1].date,
        has_data=True,
    )


# ===== 单号频率统计 =====

def _frequency(draws, getter) -> Counter:
    counter = Counter()
    for d in draws:
        counter.update(getter(d))
    return counter


def _classify_levels(stats: list) -> list:
    if not stats:
        return stats
    freqs = sorted({s.frequency for s in stats})
    n = len(freqs)
    low_threshold = freqs[max(0, n // 3) - 1] if n >= 3 else freqs[0]
    high_threshold = freqs[min(n - 1, 2 * n // 3)]
    if low_threshold == high_threshold:
        low_threshold = high_threshold - 1

    for s in stats:
        if s.frequency >= high_threshold and high_threshold > low_threshold:
            s.level = "hot"
        elif s.frequency <= low_threshold:
            s.level = "cold"
        else:
            s.level = "warm"
    return stats


def compute_number_stats(lottery_type: str):
    draws = load_history(lottery_type)
    rule = RULES[lottery_type]
    primary_pool_size = len(rule["primary_range"])
    secondary_pool_size = len(rule["secondary_range"])
    expected_primary = rule["primary_count"] / primary_pool_size
    expected_secondary = rule["secondary_count"] / secondary_pool_size

    primary_counter = _frequency(draws, lambda d: d.primary())
    secondary_counter = _frequency(draws, lambda d: d.secondary())

    total = max(1, len(draws))

    primary_stats = []
    for n in rule["primary_range"]:
        freq = primary_counter.get(n, 0)
        rate = freq / total
        primary_stats.append(NumberStat(
            number=n,
            frequency=freq,
            rate=rate,
            expected_rate=expected_primary,
            deviation=rate - expected_primary,
            level="warm",
        ))

    secondary_stats = []
    for n in rule["secondary_range"]:
        freq = secondary_counter.get(n, 0)
        rate = freq / total
        secondary_stats.append(NumberStat(
            number=n,
            frequency=freq,
            rate=rate,
            expected_rate=expected_secondary,
            deviation=rate - expected_secondary,
            level="warm",
        ))

    _classify_levels(primary_stats)
    _classify_levels(secondary_stats)

    return primary_stats, secondary_stats, len(draws)


def get_top_bottom(lottery_type: str, n: int = 5) -> dict:
    primary_stats, secondary_stats, total = compute_number_stats(lottery_type)
    if total == 0:
        return {
            "primary_hot": [], "primary_cold": [],
            "secondary_hot": [], "secondary_cold": [],
            "total_draws": 0,
        }

    p_hot = sorted(primary_stats, key=lambda s: s.frequency, reverse=True)[:n]
    p_cold = sorted(primary_stats, key=lambda s: s.frequency)[:n]
    s_hot = sorted(secondary_stats, key=lambda s: s.frequency, reverse=True)[:n]
    s_cold = sorted(secondary_stats, key=lambda s: s.frequency)[:n]

    return {
        "primary_hot": [s.to_dict() for s in p_hot],
        "primary_cold": [s.to_dict() for s in p_cold],
        "secondary_hot": [s.to_dict() for s in s_hot],
        "secondary_cold": [s.to_dict() for s in s_cold],
        "total_draws": total,
    }


# ===== 命中分析 =====

def _missing_number(n: int, pool_size: int, total: int) -> dict:
    return {
        "number": n,
        "frequency": 0,
        "rate": 0.0,
        "expected_rate": pool_size / max(1, total) if total else 0.0,
        "deviation": 0.0,
        "level": "warm",
    }


def analyze_bet(lottery_type: str, primary_numbers, secondary_numbers) -> HitAnalysis:
    primary_stats, secondary_stats, total = compute_number_stats(lottery_type)
    draws = load_history(lottery_type)

    p_set = set(primary_numbers)
    s_set = set(secondary_numbers)

    p_by_number = {s.number: s.to_dict() for s in primary_stats}
    s_by_number = {s.number: s.to_dict() for s in secondary_stats}

    primary_hits = []
    secondary_hits = []
    full_match_issues = []

    p_distribution = Counter()
    s_distribution = Counter()

    for d in draws:
        p_hits = len(p_set & d.primary_set())
        s_hits = len(s_set & d.secondary_set())
        primary_hits.append(p_hits)
        secondary_hits.append(s_hits)
        p_distribution[p_hits] += 1
        s_distribution[s_hits] += 1
        if p_hits == len(p_set) and s_hits == len(s_set):
            full_match_issues.append(d.issue)

    primary_avg = (sum(primary_hits) / len(primary_hits)) if primary_hits else 0.0
    secondary_avg = (sum(secondary_hits) / len(secondary_hits)) if secondary_hits else 0.0

    p_selected = [p_by_number.get(n, _missing_number(n, len(primary_numbers), total)) for n in primary_numbers]
    s_selected = [s_by_number.get(n, _missing_number(n, len(secondary_numbers), total)) for n in secondary_numbers]

    level_breakdown = {"hot": 0, "warm": 0, "cold": 0}
    for stat in p_selected + s_selected:
        level_breakdown[stat["level"]] = level_breakdown.get(stat["level"], 0) + 1

    bet_payload = (
        {"reds": sorted(primary_numbers), "blue": secondary_numbers[0]}
        if lottery_type == "ssq"
        else {"fronts": sorted(primary_numbers), "backs": sorted(secondary_numbers)}
    )

    return HitAnalysis(
        bet=bet_payload,
        total_draws=total,
        primary_stats=p_selected,
        secondary_stats=s_selected,
        primary_avg_hits=primary_avg,
        secondary_avg_hits=secondary_avg,
        primary_hit_distribution=dict(sorted(p_distribution.items())),
        secondary_hit_distribution=dict(sorted(s_distribution.items())),
        full_match_count=len(full_match_issues),
        full_match_issues=full_match_issues[:10],
        level_breakdown=level_breakdown,
    )


# ===== 简易 CLI =====

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="彩票历史数据统计")
    sub = parser.add_subparsers(dest="cmd")

    info_p = sub.add_parser("info", help="查看历史数据概况")
    info_p.add_argument("--type", choices=["ssq", "dlt"], default="ssq")

    top_p = sub.add_parser("top", help="查看热号 / 冷号")
    top_p.add_argument("--type", choices=["ssq", "dlt"], default="ssq")
    top_p.add_argument("-n", type=int, default=5)

    analyze_p = sub.add_parser("analyze", help="对一组号码做命中分析")
    analyze_p.add_argument("--type", choices=["ssq", "dlt"], default="ssq")
    analyze_p.add_argument("--primary", required=True)
    analyze_p.add_argument("--secondary", required=True)

    args = parser.parse_args()

    if args.cmd == "info":
        s = get_summary(args.type)
        if not s.has_data:
            print(f"{args.type} 暂无历史数据，请先运行 fetch_history.py 拉取数据。")
        else:
            print(f"彩种: {args.type}")
            print(f"总期数: {s.total_draws}")
            print(f"最早: {s.earliest_issue} ({s.earliest_date})")
            print(f"最新: {s.latest_issue} ({s.latest_date})")
    elif args.cmd == "top":
        result = get_top_bottom(args.type, args.n)
        if result["total_draws"] == 0:
            print(f"{args.type} 暂无历史数据。")
        else:
            print(f"\n=== {args.type.upper()} 热号 / 冷号 Top {args.n}（基于 {result['total_draws']} 期）===")
            for section, label in [
                ("primary_hot", "主区热号"),
                ("primary_cold", "主区冷号"),
                ("secondary_hot", "副区热号"),
                ("secondary_cold", "副区冷号"),
            ]:
                items = result[section]
                if items:
                    print(f"\n{label}: " + " ".join(
                        f"{it['number']:02d}({it['frequency']})" for it in items
                    ))
    elif args.cmd == "analyze":
        primary = [int(x) for x in args.primary.replace(",", " ").split() if x.strip()]
        secondary = [int(x) for x in args.secondary.replace(",", " ").split() if x.strip()]
        r = analyze_bet(args.type, primary, secondary)
        d = r.to_dict()
        print(f"\n=== {args.type.upper()} 命中分析（基于 {d['total_draws']} 期）===")
        print(f"主区号码: {primary} 平均每期命中: {d['primary_avg_hits']}")
        print(f"副区号码: {secondary} 平均每期命中: {d['secondary_avg_hits']}")
        print(f"主区命中分布: {d['primary_hit_distribution']}")
        print(f"副区命中分布: {d['secondary_hit_distribution']}")
        print(f"完全匹配期数: {d['full_match_count']}")
        print(f"号码热度: {d['level_breakdown']}")
    else:
        parser.print_help()
